File: scrapers/warren_nolan.py
"""
Warren Nolan scraper for college baseball predictions.
"""
import re
import requests
from bs4 import BeautifulSoup
from typing import List, Dict


def clean_team_name(name: str) -> str:
    """Clean team name by removing rankings, records, and extra whitespace."""
    # Remove rankings like "#1"
    name = re.sub(r'#\d+\s*', '', name)
    # Remove records like "(15-3)"
    name = re.sub(r'\(\d+-\d+\)', '', name)
    # Remove extra whitespace
    name = re.sub(r'\s+', ' ', name).strip()
    return name


def parse_probability(prob_text: str) -> float:
    """Parse probability from text (handles % format or decimal)."""
    prob_text = prob_text.strip()

    if '%' in prob_text:
        # Handle percentage format like "62%"
        return float(prob_text.strip('%')) / 100
    else:
        # Handle decimal format like "0.62"
        return float(prob_text)


def scrape_predictions(date_str: str) -> List[Dict]:
    """
    Scrape game predictions from Warren Nolan for a specific date.

    Args:
        date_str: Date in YYYY-MM-DD format

    Returns:
        List of game dictionaries with predictions
    """
    url = f"https://www.warrennolan.com/baseball/2024/schedules-date/{date_str}"
    headers = {
        "User-Agent": "D1BaseballPicks/1.0 (Educational Use)"
    }

    print(f"   Fetching from: {url}")

    try:
        response = requests.get(url, headers=headers, timeout=10)
        response.raise_for_status()
    except requests.RequestException as e:
        print(f"   ❌ Error fetching Warren Nolan: {e}")
        return []

    soup = BeautifulSoup(response.text, 'html.parser')
    games = []

    # Warren Nolan typically uses tables for schedule data
    # We need to find the right table - look for game data
    tables = soup.find_all('table')

    print(f"   Found {len(tables)} tables on page")

    # Try to find the schedule table
    # This may need adjustment based on actual HTML structure
    game_rows = []

    for table in tables:
        rows = table.find_all('tr')
        for row in rows:
            cells = row.find_all(['td', 'th'])
            cell_text = ' '.join([c.get_text(strip=True) for c in cells])

            # Look for rows that contain team matchups (@ or vs)
            if '@' in cell_text or ' vs ' in cell_text.lower():
                game_rows.append(row)

    print(f"   Found {len(game_rows)} potential game rows")

    for row in game_rows:
        try:
            cells = row.find_all(['td', 'th'])
            cell_texts = [c.get_text(strip=True) for c in cells]

            # Try to parse the row
            # This is a heuristic approach - may need adjustment
            teams_text = None
            prob_text = None
            time_text = None

            for text in cell_texts:
                # Look for team matchup
                if '@' in text or ' vs ' in text.lower():
                    teams_text = text
                # Look for probability (percentage or decimal between 0 and 1)
                elif '%' in text or (text.replace('.', '').isdigit() and 0 < float(text) < 1):
                    prob_text = text
                # Look for time
                elif ':' in text and any(c.isdigit() for c in text):
                    time_text = text

            if not teams_text:
                continue

            # Parse teams
            if '@' in teams_text:
                parts = teams_text.split('@')
                away_team = clean_team_name(parts[0])
                home_team = clean_team_name(parts[1])
                venue_type = 'home_b'
                team_a, team_b = away_team, home_team
                team_a_home, team_b_home = False, True
            elif ' vs ' in teams_text.lower():
                parts = re.split(r'\s+vs\s+', teams_text, flags=re.IGNORECASE)
                team_a = clean_team_name(parts[0])
                team_b = clean_team_name(parts[1])
                venue_type = 'neutral'
                team_a_home, team_b_home = False, False
            else:
                continue

            # Parse probability if available
            if prob_text:
                try:
                    prob = parse_probability(prob_text)
                    # Assume probability is for the home team or second team
                    model_prob_b = prob
                    model_prob_a = 1 - prob
                except ValueError:
                    # Skip games without valid probability
                    continue
            else:
                # Skip games without probability
                continue

            game_id = f"{date_str}_{team_a}_{team_b}".replace(' ', '_').lower()

            game = {
                'game_id': game_id,
                'date': date_str,
                'start_time': time_text or 'TBD',
                'team_a': team_a,
                'team_b': team_b,
                'team_a_home': team_a_home,
                'team_b_home': team_b_home,
                'venue_type': venue_type,
                'model_prob_a': model_prob_a,
                'model_prob_b': model_prob_b
            }

            games.append(game)
            print(f"   ✓ {team_a} vs {team_b} - {model_prob_a:.1%} / {model_prob_b:.1%}")

        except Exception as e:
            print(f"   ⚠️  Error parsing row: {e}")
            continue

    if not games:
        print("   ⚠️  No games with probabilities found.")
        print("   💡 You may need to manually inspect Warren Nolan HTML and adjust the scraper.")

    return games


if __name__ == "__main__":
    # Test scraper
    import sys
    from datetime import datetime, timedelta

    # Default to tomorrow
    tomorrow = (datetime.now() + timedelta(days=1)).strftime('%Y-%m-%d')
    date = sys.argv[1] if len(sys.argv) > 1 else tomorrow

    print(f"Testing Warren Nolan scraper for {date}...")
    games = scrape_predictions(date)
    print(f"\nFound {len(games)} games total")

File: src/lib/scrapers/warren-nolan.ts
import * as cheerio from 'cheerio';
import type { Game } from '../types';

function cleanTeamName(name: string): string {
  // Remove rankings like "#1"
  name = name.replace(/#\d+\s*/g, '');
  // Remove records like "(15-3)"
  name = name.replace(/\(\d+-\d+\)/g, '');
  // Remove extra whitespace
  name = name.replace(/\s+/g, ' ').trim();
  return name;
}

function parseProbability(probText: string): number {
  probText = probText.trim();

  if (probText.includes('%')) {
    return parseFloat(probText.replace('%', '')) / 100;
  } else {
    return parseFloat(probText);
  }
}

export async function scrapeWarrenNolan(dateStr: string): Promise<Game[]> {
  const year = dateStr.split('-')[0];
  const url = `https://www.warrennolan.com/baseball/${year}/schedules-date/${dateStr}`;

  console.log(`   Fetching from: ${url}`);

  try {
    const response = await fetch(url, {
      headers: {
        'User-Agent':
          'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36',
        Accept: 'text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8',
        'Accept-Language': 'en-US,en;q=0.5',
        'Accept-Encoding': 'gzip, deflate, br',
        DNT: '1',
        Connection: 'keep-alive',
        'Upgrade-Insecure-Requests': '1',
      },
    });

    if (!response.ok) {
      console.warn(`   ❌ Warren Nolan returned ${response.status}`);
      return [];
    }

    const html = await response.text();
    const $ = cheerio.load(html);
    const games: Game[] = [];

    // Find all tables
    const tables = $('table');
    console.log(`   Found ${tables.length} tables on page`);

    // Find game rows (rows containing @ or vs)
    const gameRows: cheerio.Cheerio<any>[] = [];

    tables.each((_, table) => {
      $(table)
        .find('tr')
        .each((_, row) => {
          const cells = $(row).find('td, th');
          const cellText = cells
            .map((_, c) => $(c).text().trim())
            .get()
            .join(' ');

          if (cellText.includes('@') || cellText.toLowerCase().includes(' vs ')) {
            gameRows.push($(row));
          }
        });
    });

    console.log(`   Found ${gameRows.length} potential game rows`);

    for (const $row of gameRows) {
      try {
        const cells = $row.find('td, th');
        const cellTexts = cells.map((_, c) => $(c).text().trim()).get();

        // Find team matchup, probability, and time
        let teamsText: string | null = null;
        let probText: string | null = null;
        let timeText: string | null = null;

        for (const text of cellTexts) {
          // Look for team matchup
          if (text.includes('@') || text.toLowerCase().includes(' vs ')) {
            teamsText = text;
          }
          // Look for probability (percentage or decimal between 0 and 1)
          else if (text.includes('%')) {
            probText = text;
          } else if (!isNaN(parseFloat(text))) {
            const val = parseFloat(text);
            if (val > 0 && val < 1) {
              probText = text;
            }
          }
          // Look for time (contains : and digits)
          else if (text.includes(':') && /\d/.test(text)) {
            timeText = text;
          }
        }

        if (!teamsText) {
          continue;
        }

        // Parse teams
        let teamA: string;
        let teamB: string;
        let venueType: 'home_a' | 'home_b' | 'neutral';
        let teamAHome: boolean;
        let teamBHome: boolean;

        if (teamsText.includes('@')) {
          const parts = teamsText.split('@');
          const awayTeam = cleanTeamName(parts[0]);
          const homeTeam = cleanTeamName(parts[1]);
          venueType = 'home_b';
          teamA = awayTeam;
          teamB = homeTeam;
          teamAHome = false;
          teamBHome = true;
        } else if (teamsText.toLowerCase().includes(' vs ')) {
          const parts = teamsText.split(/\s+vs\s+/i);
          teamA = cleanTeamName(parts[0]);
          teamB = cleanTeamName(parts[1]);
          venueType = 'neutral';
          teamAHome = false;
          teamBHome = false;
        } else {
          continue;
        }

        // Parse probability if available
        let modelProbA: number;
        let modelProbB: number;

        if (probText) {
          try {
            const prob = parseProbability(probText);
            // Assume probability is for team B (home team or second team)
            modelProbB = prob;
            modelProbA = 1 - prob;
          } catch {
            continue;
          }
        } else {
          continue;
        }

        const gameId = `${dateStr}_${teamA}_${teamB}`.replace(/\s+/g, '_').toLowerCase();

        const game: Game = {
          id: gameId,
          date: dateStr,
          startTime: timeText || 'TBD',
          teamA: teamA.toLowerCase(),
          teamB: teamB.toLowerCase(),
          teamAHome,
          teamBHome,
          venueType,
          modelProbA,
          modelProbB,
        };

        games.push(game);
        console.log(
          `   ✓ ${teamA} vs ${teamB} - ${(modelProbA * 100).toFixed(1)}% / ${(modelProbB * 100).toFixed(1)}%`
        );
      } catch (error) {
        console.warn(`   ⚠️  Error parsing row: ${error}`);
        continue;
      }
    }

    if (games.length === 0) {
      console.log('   ⚠️  No games with probabilities found.');
      console.log('   💡 You may need to manually inspect Warren Nolan HTML and adjust the scraper.');
    }

    return games;
  } catch (error) {
    console.error(`   ❌ Error fetching Warren Nolan: ${error}`);
    return [];
  }
}
